### PR TITLE
Re-architecture flash for fast payments

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
No need for a review yet, it needs a cleanup. It's a CI test for now. It does work great on my local machine though.

Also includes https://github.com/bpfkorea/agora/pull/1673 in the first commit.

Fixes https://github.com/bpfkorea/agora/issues/1644

### CODE REVIEW

File: source/agora/flash/UpdateSigner.d [hunk 1]
```diff
@@ -273,16 +297,23 @@ public class UpdateSigner
...
 
+        writefln("%s: Settlement signature %s from %s received",
+            this.kp.V.prettify, seq_id, this.peer_pk.prettify);
+
         if (auto error = this.isInvalidSettleMultiSig(this.pending_settle,
             settle_res.value, priv_nonce, peer_nonce))
         {
             // todo: inform? ban?
-            writefln("%s: Error during validation: %s. For settle signature "
+            writefln("%s: Error durin g validation: %s. For settle signature "
```
Comment (kchulj) [inline source/agora/flash/UpdateSigner.d]: I think this is a typo.

Comment (AndrejMitrovic) [reply source/agora/flash/UpdateSigner.d]: Yeah there's going to be lots of these here. But thanks.

**Comment by AndrejMitrovic** [comment]
I can reproduce the thread bug where it's getting stuck on every run locally: https://github.com/AndrejMitrovic/agora/tree/thread-bug

Ran with: `dchatty=true dnologs=true dtest=agora.test.Flash dub test`.

Tried with latest ldc as well. It seems all the threads are waiting for something.

**Comment by AndrejMitrovic** [comment]
Edit: removed old comment. I think the stack trace was deceiving.

**Comment by AndrejMitrovic** [comment]
✅

**Comment by AndrejMitrovic** [comment]
I think I might have reverted some of the error reporting routines. But I don't think they were designed right. The error reporting for a failed payment should be done through another API, not through a series of blocking calls.

**Comment by omerfirmak** [comment]
> I think I might have reverted some of the error reporting routines. But I don't think they were designed right. The error reporting for a failed payment should be done through another API, not through a series of blocking calls.

IIRC, they werent blocking.

**Comment by AndrejMitrovic** [comment]
> IIRC, they werent blocking.

Ah hold on, I think I see it now. I'll revert some of my changes.

**Comment by AndrejMitrovic** [comment]
I re-added error reporting. I will write a separate test-case for this later though, as I don't want to further delay the PR.

**Comment by AndrejMitrovic** [comment]
I'm gonna go ahead with this so we can continue rapid development.

**Pull request merged**